findthemasks/findthemasks
# Map search functionality (search by zip, address, etc). Take 2!

Fixes issue #115. This is the second attempt at implementing a fix, first PR was #192 (was scrapped due to modifying too much and subsequent conflicts introduced by #221). In this revision, I've kept ALL modifications of existing code down to the bare minimum in order to reduce merge conflicts. I've compensated by including a lot of suggestions inline via `TODO`'s to flag areas which are either duplicating or adjacent to map search functionality that I think could use improvement. These are also documented in "Recommendations" below.

cc reviewers and lead devs @graemer and @awong-dev
cc stakeholders @r-pop and @Shuhanhe1

**Preview links:**
- https://findthemasks-issue-115-ctl5pjq.herokuapp.com/give.html?show-search=true
- https://findthemasks-issue-115-ctl5pjq.herokuapp.com/index.html?show-search=true

![take 2, demo 2](https://user-images.githubusercontent.com/4269377/77710433-31829480-6f8b-11ea-999e-925a2d256a89.gif)

**New features:**

- **Disabled by default** (during BETA period).
- Ability to search by arbitrary user entered location, e.g. zip code, address and etc.
- Ability to use current location **Requires SSL, see "Testing locally" below.**
- New parameter options:
  - `hide-search=true` to initialize **without** the map search / autosuggest field. Only relevant once this is enabled by default.
  - `show-search=true` to enable the the map search / autosuggest field during the beta period.
- Adding `data-enable-search` atribute (can be either `true` or `false`) to allow default enabling/disabling of the search field by default on a **per-page basis**.

**Notes:**

- Let me know if we should initialize by default without autosuggest. This has implications on https://getusppe.org/give/ which has an `<iframe>` that initializes with limited height (relates to issue #170)
- I made sure that this implemented `i18n`, but translations from Google Translations may be inaccurate and there may be missing languages by the time this is merged.

**Testing locally:**

1. Clone `git@github.com:patricknelson/findthemasks.git`
1. Checkout `issue-115-map-search-take-2`
1. Start Chrome with the flag `--unsafely-treat-insecure-origin-as-secure="http://local.findthemasks.com/"` (this will allow geocoding on insecure connection)
1. Load this URL (the maps key will automatically be set):
http://local.findthemasks.com/give.html?mapsKey=AIzaSyCVh7x85vFsElLwVQ2M18PtJynLOQ_E7Gw

Recommendations **once merged** (see `TODO`'s for details):

- Reformat file indentation to project standard (e.g. `addMarkerToMap()` needs work at time of writing).
- Simplify function arity, e.g. reference global `map` and `markers` variables instead of passing to functions.
- Move to simplified interface for recentering/zooming map, implemented in this PR, specifically the primary function  `centerMapToMarkersNearCoords(latitude, longitude)` and the helper `centerMapToMarkersNearUser()` (which _should_ be called on startup instead of the harder to call and understand `centerMapToNearestMarkers(map, markers, bounds)`). This allowed me to easily position map at user coordinates or at geocoded coordinates from arbitrary user input (search field) with a very simple standardized interface.
- **showMarkers():** Standardize `showMarkers()` to be responsible for only showing markers; currently does too much (including zooming/centering map via deeper calls).
- **initMap():** Consider moving centering/zooming of map also out of `initMap()` so that it's only responsible for initializing the map and then takes a callback to allow further config/modification once done, or use a global (module level) config that init map depends on to know how to initialize center/zoom/filters (reason: so "Reset map" link can also rely on this).
